index.js

```diff
@@ -37,35 +38,31 @@ const createRetrieval = async (res, client) => {
   })
 }
 
-const updateRetrieval = async (req, res, client, retrievalId) => {
+const setRetrievalResult = async (req, res, client, retrievalId) => {
   assert(!Number.isNaN(retrievalId), 400, 'Invalid Retrieval ID')
   const body = await getRawBody(req, { limit: '100kb' })
-  const { success, walletAddress } = JSON.parse(body)
-  assert.strictEqual(
-    typeof success,
-    'boolean',
-    400,
-    'boolean .success required'
-  )
-  assert.strictEqual(
-    typeof walletAddress,
-    'string',
-    400,
-    'string .walletAddress required'
-  )
-  const { rows } = await client.query(`
-    UPDATE retrievals
-    SET finished_at = NOW(),
-      success = $2,
-      wallet_address = $3
-    WHERE id = $1 AND success IS NULL
-    RETURNING id
-  `, [
-    retrievalId,
-    success,
-    walletAddress
-  ])
-  assert(rows.length > 0, 404, 'Retrieval Not Found')
+  const result = JSON.parse(body)
+  try {
+    await client.query(`
+      INSERT INTO retrieval_results
+      (retrieval_id, wallet_address, success)
+      VALUES ($1, $2, $3)
+    `, [
+      retrievalId,
+      result.walletAddress,
+      result.success
+    ])
+  } catch (err) {
+    if (err.constraint === 'retrieval_results_retrieval_id_fkey') {
+      assert.fail(404, 'Retrieval Not Found')
+    } else if (err.constraint === 'retrieval_results_retrieval_id_key') {
+      assert.fail(409, 'Retrieval Already Completed')
```

Comment:
hold on, this approach might not scale well enough :( Working on another PR, I just discovered a validation error where nothing useful is on the error object. So, I guess we mirror validations in JS?

Reply:
see the update in https://github.com/filecoin-station/spark-api/pull/20